### ISSUE TITLE
Cannot build --profile APK for a plugin with a custom dependency added to 'debugImplementation' and 'releaseImplementation'

### ISSUE THREAD

**Issue by wujek-srujek** [flutter/flutter#133931]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create a plugin project with Android as one of the target platforms.
2. Add a custom dependency to `debuIimplementation` and `releaseImplementation`, and make sure the APKs can be built.
3. Try to build `flutter build apk --profile` - this will fail with compilation error that it cannot find the types that come from the custom library.

Invoking `flutter build apk --release` and `--debug` works fine. I tried adding the dependency to `profileImplementation`, but such a thing doesn't even exist and Gradle fails.

How do I add a custom dependency to build for `--profile`?

### Expected results

A profile APK can be created.

### Actual results

The build fails with compilation errors saying types from the custom library cannot be resolved. This means that the 'profile' classpath isn't using the custom dependency.

### Code sample

<details><summary>Code sample</summary>

<plugin root>/android/build.gradle
```groovy
...
debugImplementation 'com.my:dep-debug:1.0.0'
releaseImplementation 'com.my:dep-release:1.0.0'
...
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.10.6, on macOS 13.5.1 22G90 darwin-arm64, locale en-DE)
    • Flutter version 3.10.6 on channel stable at /Users/rswierzy/fvm/versions/3.10.6
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f468f3366c (7 weeks ago), 2023-07-12 15:19:05 -0700
    • Engine revision cdbeda788a
    • Dart version 3.0.6
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/rswierzy/Library/Android/sdk
    • Platform android-33, build-tools 33.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.70.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.5.1 22G90 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 116.0.5845.140

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
Thanks for the report. I am not sure about `custom dependency to build for --profile` and how to do it for `--profile` mode.
I'll keep this issue open for team's input / attention.

**Comment by gmackall** [comment]
@wujek-srujek can you provide the compilation error you are seeing?

**Comment by wujek-srujek** [comment]
Here you go. Please forgive me but I needed to 'de-proprietatize' the paths, types and so on. I also included 2 of the compilation errors I'm seeing, no reason to include all 150 or so, they are all errors that members of the custom dependency cannot be resolved.

It seems like the 'profile' Gradle configuration doesn't include either 'debug' or 'release' and the library simply isn't on the classpath. But there is no 'profileImplementation' that I could use to make it work.

```shell
$ flutter build apk --release

Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 1396 bytes (99.9% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when building your app.
Running Gradle task 'assembleRelease'...                           32.5s
✓  Built build/app/outputs/flutter-apk/app-release.apk (19.1MB).
feat/my-branch example $ flutter build apk --profile
feat/my-branch example $ flutter build apk --profile

Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 1396 bytes (99.9% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when building your app.
e: file:///Users/wujek/Development/myplugin/android/src/main/kotlin/com/foo/bar/Baz.kt:6:20 Unresolved reference: apackagefromimportedlibrary
e: file:///Users/wujek/Development/myplugin/android/src/main/kotlin/com/foo/bar/Baz.kt:28:5 'logMessage' overrides nothing
// 150 other similar errors, all of which mention a package, type, method, property etc. from imported library.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':rclib:compileProfileKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 24s
Running Gradle task 'assembleProfile'...                           24.8s
Gradle task assembleProfile failed with exit code 1
```

BTW, I'm doing it on Flutter 3.13.3 right now.

**Comment by gmackall** [comment]
Potentially related: https://github.com/flutter/flutter/issues/48041#issuecomment-570721663, it looks like this may have come up in the past and this comment has a way that might force the profile implementation to include `debug`?

**Comment by wujek-srujek** [comment]
Looks like my ticket is a duplicate of an issue that was never really fixed. The solution from the comment will of course work, this is a workaround that I also came up with, but wouldn't it be nicer if Futter provided a solution to build profile, instead of everybody relying on such custom solutions? At the same time, it does seem like a niche use case as it pops up every 3 years...

As for the workaround itself - I was always under the impression that a profile build is more a 'release build + some extra stuff for profiling' than a 'debug build with said extra stuff'. As such, shouldn't the profile configuration be initialized with release rather than debug? I also think adding a dedicated 'profileImplementation' configuration would be much more explicit and less magical. This is BTW something that doesn't work for me as of now, contrary to what https://github.com/flutter/flutter/issues/48041#issuecomment-570762118 alleges.

**Comment by gmackall** [comment]
> Looks like my ticket is a duplicate of an issue that was never really fixed. The solution from the comment will of course work, this is a workaround that I also came up with, but wouldn't it be nicer if Flutter provided a solution to build profile, instead of everybody relying on such custom solutions? At the same time, it does seem like a niche use case as it pops up every 3 years...

Yes, I agree it would be better if the workaround wasn't required. And yeah it is only that - a workaround. An actual solution would definitely be better, it's just that this issue will probably not be at the top of the list to pick up for the time being.

> As for the workaround itself - I was always under the impression that a profile build is more a 'release build + some extra stuff for profiling' than a 'debug build with said extra stuff'. As such, shouldn't the profile configuration be initialized with release rather than debug? I also think adding a dedicated 'profileImplementation' configuration would be much more explicit and less magical. This is BTW something that doesn't work for me as of now, contrary to what [#48041 (comment)](https://github.com/flutter/flutter/issues/48041#issuecomment-570762118) alleges.

I agree with that characterization of a profile build, it probably would make more sense to be initialized as release.
